core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java

```diff
@@ -334,48 +330,59 @@ protected Queue<Node> maybeAddDcFailover(@Nullable Request request, @NonNull Que
-                liveNodes.dcs().stream()
-                    .filter(Predicates.not(Predicates.equalTo(localDc)))
-                    .sorted(
-                        Comparator.comparingInt(
-                            dc -> {
-                              int i = preferredRemoteDcs.indexOf(dc);
-                              return i < 0 ? Integer.MAX_VALUE : i;
-                            }))
-                    .flatMap(
-                        dc -> {
-                          if (preferredRemoteDcs.isEmpty()) {
-                            return liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc);
-                          } else {
-                            final Object[] nodesPerRemoteDc =
-                                liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc).toArray();
-                            if (nodesPerRemoteDc.length > 0) {
-                              shuffleHead(nodesPerRemoteDc, nodesPerRemoteDc.length);
-                            }
-                            return Stream.of(nodesPerRemoteDc);
-                          }
-                        })
-                    .toArray();
-
-            int remoteNodesLength = remoteNodes.length;
-            if (remoteNodesLength == 0) {
-              return EMPTY_NODES;
-            }
-
-            if (preferredRemoteDcs.isEmpty()) {
-              shuffleHead(remoteNodes, remoteNodesLength);
-            }
-
-            return remoteNodes;
-          }
-        };
-
-    return new CompositeQueryPlan(local, remote);
+    if (preferredRemoteDcs.isEmpty()) {
+      return new CompositeQueryPlan(local, buildRemoteQueryPlanAll());
+    }
+    return new CompositeQueryPlan(local, buildRemoteQueryPlanPreferred());
+  }
+
+  private QueryPlan buildRemoteQueryPlanAll() {
+
+    return new LazyQueryPlan() {
+      @Override
+      protected Object[] computeNodes() {
+
+        Object[] remoteNodes =
+            liveNodes.dcs().stream()
+                .filter(Predicates.not(Predicates.equalTo(localDc)))
+                .flatMap(dc -> liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc))
+                .toArray();
+        if (remoteNodes.length == 0) {
+          return EMPTY_NODES;
+        }
+        shuffleHead(remoteNodes, remoteNodes.length);
+        return remoteNodes;
+      }
+    };
+  }
+
+  private QueryPlan buildRemoteQueryPlanPreferred() {
+
+    Set<String> dcs = liveNodes.dcs();
+    List<String> orderedDcs = Lists.newArrayListWithCapacity(dcs.size());
+    orderedDcs.addAll(preferredRemoteDcs);
+    orderedDcs.addAll(Sets.difference(dcs, preferredRemoteDcs));
+
+    QueryPlan[] queryPlans =
+        orderedDcs.stream()
+            .filter(Predicates.not(Predicates.equalTo(localDc)))
+            .map(
+                (dc) -> {
+                  return new LazyQueryPlan() {
+                    @Override
+                    protected Object[] computeNodes() {
+                      Object[] rv = liveNodes.dc(dc).toArray();
```

Comment:
Should this include `maxNodesPerRemoteDc` like:                       
`Object[] rv = liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc).toArray();
`

Reply:
Good catch @nitinitt, you're absolutely right!

Updating.